include/swift/AST/DiagnosticEngine.h

```diff
@@ -515,6 +528,42 @@ namespace swift {
     /// emitted as a warning, but a note will still be emitted as a note.
     InFlightDiagnostic &limitBehavior(DiagnosticBehavior limit);
 
+    /// Wraps this diagnostic in another diagnostic. That is, \p wrapper will be
+    /// emitted in place of the diagnostic that otherwise would have been
+    /// emitted.
+    ///
+    /// The first argument of \p wrapper must be of type 'Diagnostic *'.
+    ///
+    /// The emitted diagnostic will have:
+    ///
+    /// \li The ID, message, and behavior of \c wrapper.
+    /// \li The arguments of \c wrapper, with the last argument replaced by the
+    ///     diagnostic currently in \c *this.
+    /// \li The location, ranges, decl, fix-its, and behavior limit of the
+    ///     diagnostic currently in \c *this.
+    InFlightDiagnostic &wrapIn(const Diagnostic &wrapper);
```

Comment:
Can you tell me more about the advantage of wrapping this diagnostic in the Xcode display? Should we apply the same trick to more kind of diagnostics or is this specific to access notes?

Reply:
Basically, Xcode is not very good at displaying notes in its UI. Wrapping a diagnostic produces a longer message, but it's a *single* message that will actually be displayed.

I've been thinking that it might be good to turn `diag::module_interface_build_failed` and friends into wrapping diagnostics around the underlying errors, since the underlying errors often seem to go unnoticed by users, but are usually a better way to resolve the problem. But that's a different pull request.